src/Nethermind/Nethermind.Network/ForkInfo.cs

```diff
@@ -70,5 +76,151 @@ public ForkId GetForkId(long headNumber, ulong headTimestamp)
 
         private static int CompareTransitionOnActivation(ForkActivation activation, (ForkActivation Activation, ForkId _) transition) =>
             activation.CompareTo(transition.Activation);
+
+        /// <summary>
+        /// Verify that the forkid from peer matches our forks. Code is largely copied from Geth.
+        /// </summary>
+        /// <param name="peerId"></param>
+        /// <returns></returns>
+        public IForkInfo.ValidationResult ValidateForkId(ForkId peerId)
+        {
+            // Run the fork checksum validation ruleset:
+            //   1. If local and remote FORK_CSUM matches, compare local head to FORK_NEXT.
+            //        The two nodes are in the same fork state currently. They might know
+            //        of differing future forks, but that's not relevant until the fork
+            //        triggers (might be postponed, nodes might be updated to match).
+            //      1a. A remotely announced but remotely not passed block is already passed
+            //          locally, disconnect, since the chains are incompatible.
+            //      1b. No remotely announced fork; or not yet passed locally, connect.
+            //   2. If the remote FORK_CSUM is a subset of the local past forks and the
+            //      remote FORK_NEXT matches with the locally following fork block number,
+            //      connect.
+            //        Remote node is currently syncing. It might eventually diverge from
+            //        us, but at this current point in time we don't have enough information.
+            //   3. If the remote FORK_CSUM is a superset of the local past forks and can
+            //      be completed with locally known future forks, connect.
+            //        Local node is currently syncing. It might eventually diverge from
+            //        the remote, but at this current point in time we don't have enough
+            //        information.
+            //   4. Reject in all other cases.
+            BlockHeader? head = _blockTree.Head?.Header;
+            if (head == null) return IForkInfo.ValidationResult.Valid;
+
+            for (int i = 0; i < Forks.Length; i++)
+            {
+                (ForkActivation forkActivation, ForkId forkId) = Forks[i];
+                bool usingTimestamp = forkActivation.Timestamp != null;
+                ulong headActivation = (usingTimestamp ? head.Timestamp : (ulong)head.Number);
```

Comment:
Did you double check this with geth's edge case? https://github.com/ethereum/go-ethereum/pull/25878/files#diff-77f77794b42f4fd84db10f281103146d1f12e9b50a404254bc87565e3e423872R193. 

Did you also copy the new test case? Coz I did not.

Reply:
Now it is. check it out @asdacap